utils/src/main/scala/za/co/absa/enceladus/utils/fs/S3FsUtils.scala

```diff
@@ -0,0 +1,271 @@
+      .delete(Delete.builder().objects(objIds.asJava).build())
+      .build()
+
+    val delResp: DeleteObjectsResponse = s3Client.deleteObjects(request)
+
+    if (delResp.errors().size() > 0) {
+      log.warn(s"Errors while deleting (${delResp.errors.size}):\n ${delResp.errors.asScala.map(_.message()).mkString("\n")}")
+    }
+  }
+
+  /**
+   * Finds the latest version given a publish folder on distributed-FS
+   *
+   * @param publishPath The distributed-FS path to the publish folder containing versions
+   * @param reportDate  The string representation of the report date used to infer the latest version
+   * @return the latest version or 0 in case no versions exist
+   */
+  override def getLatestVersion(publishPath: String, reportDate: String): Int = {
+
+    // setup accumulation
+    val location = publishPath.toS3Location(region)
+    val initVersion = 0
+
+    // looking for $publishPath/enceladus_info_date=$reportDate\enceladus_info_version=$version
+    val prefix = s"${location.path}/enceladus_info_date=$reportDate/enceladus_info_version="
+    val prefixedLocation = location.copy(path = prefix)
+
+    def accumulateSizeOp(previousMaxVersion: Int, response: ListObjectsV2Response): Int = {
+      val objects = response.contents().asScala
+
+      val existingVersions = objects
+        .map(_.key)
+        .flatMap { key =>
+          assert(key.startsWith(prefix), s"Retrieved keys should start with $prefix, but precondition fails for $key")
+          val noPrefix = key.stripPrefix(prefix)
+          Try {
+            noPrefix.takeWhile(_.isDigit).toInt // may not hold valid int >= 1
+          } match {
+            case Success(version) if version >= 1 => Some(version)
+            case _ => None
+          }
+        }
+        .toSet
+
+      if (existingVersions.isEmpty) {
+        previousMaxVersion
+      } else {
+        Math.max(previousMaxVersion, existingVersions.max)
+      }
+    }
+
+    listAndAccumulateRecursively(prefixedLocation, accumulateSizeOp, initVersion)
+  }
+
+  private[fs] def getS3Client: S3Client = S3Utils.getS3Client(region, credentialsProvider)
+
+  /**
+   * General method to list and accumulate the objects info. Note, that the method strives to be memory-efficient -
+   * i.e. accumulate the current batch first and then load the next batch (instead of the naive "load all first, process later"
+   *
+   * @param location        s3location - bucket & path are used
+   * @param accumulateOp    operation to accumulate
+   * @param initialAccValue (initial/carry-over) accumulator value
+   * @param breakOut        allows to break the recursion prematurely when the defined value equals the currently accumulated value.
+   *                        Default: None = no break out
+   * @tparam T accumulator value type
+   * @return accumulated value
+   */
+  private def listAndAccumulateRecursively[T](location: S3Location,
```

Comment:
All List-and-do-something methods are using this helper in order to support pagination > 1000 items. The related test suite attempts to test this property with an optional breakOut (early recursion return).